Livemode: various improvements to make the feature more accessible to users

MichalHe opened the pull request:
## Livemode improvements
This is a first part of the work necessary to make the  `livemode` feature more accessible, hoping to make it the default -in the future. There are multiple quality-of-live implemented as part of this PR (kept as separate commits):
- CLI: introduce `--enable-experimental-feature` that allows users to simply enable entire group of actors
- exclude DNF cache from squashfs to reduce the size of the created image. The DNF cache that contains packages to be installed during the upgrade is always read from `/sysroot/var/lib/leapp`, and therefore, it would not be read from squashfs'ed `/var/lib/leapp`
- configuration: use configuration capabilities provided by the framework instead of ad-hoc INI file
- bugfix: do not crash with an unhandled exception when trying to enable a systemd service inside the squashfs image that already appears to be enabled

## Why are we introducing a new command line option `--enable-experimental-feature`?
1) `--whitelist-experimental` is a quite bad user experience. Why should a user know actor names? The user is more interested in functionality, not in leapp-repository's internals.
2) Some features like `livemode` use a bunch of actors to facilitate the feature, therefore, the user would need to list all of these actors explicitly and he/she better not forget any of them.

Why don't we just use a single experimental actor that produces a message which will "enable" (if the message is received) all of the remaining actors? The problem is that anyone can produce such a message. We could write the remaining experimental actors such that they check for a combination of receiving a special message + checking that `LEAPP_EXPERIMENTAL` is set, but this solution is cumbersome and far from perfect (user can enable a different experimental actor, and, unintentionally enable livemode if some actor is producing the special message). Finally, there is no way for actors to check whether a specific experimental actor has been enabled, if I am not wrong -- `LEAPP_EXPERIMENTAL` is binary information (yes/no). 

## Example of livemode configuration
```
livemode:
  additional_packages : [ vim ]
  autostart_upgrade_after_reboot : false
  setup_network_manager : true
  setup_passwordless_root : true
  setup_opensshd_using_auth_keys : /root/.ssh/authorized_keys
```

jira-ref: RHELMISC-10648

Commit:
cli(upgrade): allow users to enable entire experimental features

Introduce a new CLI option --enable-experimental-feature that allows
users to enable entire features that might be facilitated by a large
number of experimental actors. Previously, the used had to
remember/figure out the names of all of these actors and list them
manually using `--whitelist-experimental`. Using
`--enable-experimental-feature` therefore lifts this burden from the
user, and the user simply needs to know what experimental feature to
enable. The help for the new options includes a list of all supported
experimental feature - at the moment, the list contains only 'livemode'.

Jira-ref: RHELMISC-10648

MichalHe commented:
## Why are we introducing a new command line option `--enable-experimental-feature`?
1) `--whitelist-experimental` is a quite bad user experience. Why should a user know actor names? The user is more interested in functionality, not in leapp-repository's internals.
2) Some features like `livemode` use a bunch of actors to facilitate the feature, therefore, the user would need to list all of these actors explicitly and he/she better not forget any of them.

Why don't we just use a single experimental actor that produces a message which will "enable" (if the message is received) all of the remaining actors? The problem is that anyone can produce such a message. We could write the remaining experimental actors such that they check for a combination of receiving a special message + checking that `LEAPP_EXPERIMENTAL` is set, but this solution is cumbersome and far from perfect (user can enable a different experimental actor, and, unintentionally enable livemode if some actor is producing the special message). Finally, there is no way for actors to check whether a specific experimental actor has been enabled, if I am not wrong -- `LEAPP_EXPERIMENTAL` is binary information (yes/no).

Commit:
feat(livemode): exclude DNF cache from the created squashfs image

Exclude /var/cache/dnf from the generated squashfs image. The DNF
cache is not needed for the live system that we boot into, since leapp
will always use DNF cache stored within target userspace container
(/sysroot/var/lib/leapp/...). Therefore, this optimization saves
a lot of disk space for the user (2GB->700mb).

Commit:
livemode(cfg): add declaration of livemode config fields

Add definitions of all config classes that describe configuration of the
livemode feature using the configurability provided by the leapp
framework. The list of configuration options remains unchanged (except
for the `is_enabled` field that is removed) when compared to the
current implementation that relies on an ad-hoc INI file. The next step
is to drop the INI-based implementation in favour of using
framework-based configs relying on the field definitions from this PR.

Jira-ref: RHELMISC-10648

Commit:
livemode(cfg): use framework's configurability instead of ad-hoc INI

Drop the ad-hoc INI-based config for livemode in favour and replace it
with configuration facilitated by the leapp framework. The list of
configuration options remains (and their semantics) remains almost
unchanged, save for some field names that are renamed in a way that
better reveals their effect to the user.

Jira-ref: RHELMISC-10648

MichalHe commented:
## Consuming user-specified configuration
### No config provided (i.e., defaults)
Produced livemode config without modifying configuration:
```
{
    "additional_packages": [],
    "autostart_upgrade_after_reboot": true,
    "capture_upgrade_strace_into": "",
    "dracut_network": "",
    "is_enabled": true,
    "setup_network_manager": false,
    "setup_opensshd_with_auth_keys": "",
    "setup_passwordless_root": false,
    "squashfs_fullpath": "/var/lib/leapp/live-upgrade.img",
    "url_to_load_squashfs_from": ""
}
```
### Modified configuration 
Using the following configuration:
```
livemode:
  additional_packages : [ vim ]
  autostart_upgrade_after_reboot : false
  setup_network_manager : true
  setup_passwordless_root : true
  setup_opensshd_using_auth_keys : /root/.ssh/authorized_keys
```
Resulted in producing the following configuration message:
```
{
    "additional_packages": [
        "vim"
    ],
    "autostart_upgrade_after_reboot": false,
    "capture_upgrade_strace_into": "",
    "dracut_network": "",
    "is_enabled": true,
    "setup_network_manager": true,
    "setup_opensshd_with_auth_keys": "/root/.ssh/authorized_keys",
    "setup_passwordless_root": true,
    "squashfs_fullpath": "/var/lib/leapp/live-upgrade.img",
    "url_to_load_squashfs_from": ""
}
```

Commit:
fix(livemode): do not stop if dbus already appears to be enabled

Some of the systemd units might be already enabled in the target
userspace container, causing an unhandled FileAlreadyExists error
when we attempt to enable them. This commit ignores such errors, working
on under the assumption that the services we wanted to enable are
already enabled. Hence, we ignore the possibility that the file which
unexpectedly resides at the destination of the symlink which enables the
service might contain some unexpected/incorrect content.

pirat89 commented:
/packit copr-build

pirat89 commented on commands/upgrade/util.py:
```diff
@@ -184,11 +196,20 @@ def handle_output_level(args):
         os.environ['LEAPP_EXPERIMENTAL'] = '0'
+        args.whitelist_experimental = []
+
+    for experimental_feature in set(args.enable_experimental_feature):
+        actors_needed_for_feature = EXPERIMENTAL_FEATURES[experimental_feature]
+        args.whitelist_experimental.extend(actors_needed_for_feature)
+    if args.enable_experimental_feature:
+        os.environ['LEAPP_EXPERIMENTAL'] = '1'
```
@MichalHe  I am thinking about the situation when there will be no experimental feature at all - which is actually possible state in future. In such a case, there could be any value set as for the option it would lead to: `choices=[]` which would mean that any value would be possible to set based on the current framework behaviour (note I give it a thought to allow to set empty choices, but the behaviour then would be kind of problematic from my POV, leaving up to discussion a possible change in the framework).

In such a case `EXPERIMENTAL_FEATURES[experimental_feature]` would end with `KeyError` So it would be better to use `get(experimental_feature, None)` and add another condition to check whether it's `None` before trying to modify `whitelist_experimental`.

I have been thinking how to deal properly with LEAPP_EXPERIMENTAL in such a case when a nonexisting value is present, as it would lead to actually no experimental code to be executed anyway when `whitelist_experimental` list is not extended. First, I thought that possibly we should not set the value to `'1'` in such a case. However, people know they are already asking for an experimental content when they use the option at the first place, so let's stick with the current condition when speaking about the `LEAPP_EXPERIMENTAL` envar.

MichalHe replied:
I have extended to code to first check whether `experimental_feature` is present in `EXPERIMENTAL_FEATURES` (if not, we skip processing it).

pirat89 reviewed (changes requested):
See my comments below for more details. Also
* the orig config file is not removed (should be the new one part of the installed rpm or we will keep on user to create it?...)

* the config scanner should be possibly updated? or the logic around the `--enable-experimental-feature` as it still requires set envar `LEAPP_DEVEL_ENABLE_LIVE_MODE=1`?  Or is the requirement of this envar still needed? See the `should_scan_config` function. From the description and name of the option (and PR description) I would expect the envar would not be needed anymore. Let's sync about it yet. Some additional info:  

cmd:
```
 leapp upgrade --debug --no-rhsm --enable-experimental-feature livemode
```

result:
```
Actor: live_mode_config_scanner
Executed: True
Phase: FactsCollection
Started: 2025-04-06T19:19:15.541892Z
Produced messages:
    ----
Executed shell commands:
    ----
Logs:
--- {"level": "DEBUG", "message": "2025-04-06 21:19:15.573 DEBUG    PID: 93812 leapp.workflow.FactsCollection.live_mode_config_scanner: Will not scan livemode config - the live mode is not enabled."}
```

Commit:
feat(livemode): remove the use of LEAPP_DEVEL_ENABLE_LIVEMODE

The environmental variable has been introduced to prevent accidental
execution of livemode. However, in order for users to use the feature,
this environmental variable introduced unnecessary friction. Therefore,
this patch removes the use of the variable. Instead, whitelisting
experimental actors that facilitate livemode should be the only
mechanism that is used to enabled/disable the feature.

Jira-ref: RHELMISC-10648

Commit:
spec: require leapp-framework 6.1 for default CLI vals

Bump framework version as we want to use default=[] for the newly
introduced `--enable-experimental-features` switch.

Jira-ref: RHELMISC-10648

MichalHe commented on commands/preupgrade/__init__.py:
```diff
@@ -14,6 +14,8 @@
 @command('preupgrade', help='Generate preupgrade report')
 @command_opt('whitelist-experimental', action='append', metavar='ActorName', help='Enables experimental actors')
+@command_opt('enable-experimental-feature', action='append', metavar='Feature',
+             help='Enable experimental feature', choices=list(util.EXPERIMENTAL_FEATURES))
```
I have updated the help message to contain a list of available experimental features. If there are no experimental features available, the `--help` contains:
```
  --enable-experimental-feature Feature
                        Enable experimental feature. Available experimental
                        features: There are no experimental features available
```
If there are some experimental features (such as livemode) we get:
```
  --enable-experimental-feature Feature
                        Enable experimental feature. Available experimental
                        features: livemode
```

Furthermore, I have verified that running `leapp upgrade --enable-experimental-feature xyz` does not produce any errors when there are no experimental features available.

pirat89 reviewed (approved):
thanks! lgtm and so far it seems working based on manual testing

note: failed IPU 7 -> 8 tests are irrelevant - it's already in the removal process
